File: bindings/rust/src/safe_json.rs
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>

//! Safe JSON validation via libproven FFI.
//!
//! Validates JSON strings and determines root value types without
//! full parsing. All operations delegate to Idris 2 verified code.

use crate::core::{self, Result};
use crate::ffi;

/// Re-export JsonType from FFI for public use.
pub use crate::ffi::JsonType;

/// Safe JSON operations.
pub struct SafeJson;

impl SafeJson {
    /// Check if a string is valid JSON.
    ///
    /// Returns `true` if the input is syntactically valid JSON.
    pub fn is_valid(input: &str) -> Result<bool> {
        let bytes = input.as_bytes();
        // SAFETY: We pass a valid pointer and length from the string's bytes.
        let result = unsafe {
            ffi::proven_json_is_valid(bytes.as_ptr(), bytes.len())
        };
        core::bool_result_to_result(result)
    }

    /// Get the JSON value type at the root level.
    ///
    /// Returns `JsonType::Invalid` if the input is not valid JSON.
    pub fn get_type(input: &str) -> JsonType {
        let bytes = input.as_bytes();
        // SAFETY: We pass a valid pointer and length from the string's bytes.
        unsafe { ffi::proven_json_get_type(bytes.as_ptr(), bytes.len()) }
    }
}

File: bindings/zig/src/safe_json.zig
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>
//
// Proven SafeJSON - FFI bindings to libproven JSON validation.
// All computation is performed in verified Idris 2 code via libproven.

const std = @import("std");
const c = @cImport(@cInclude("proven.h"));

/// JSON value types (matching ProvenJsonType enum in proven.h).
pub const JsonType = enum(i32) {
    null_type = c.PROVEN_JSON_NULL,
    bool_type = c.PROVEN_JSON_BOOL,
    number = c.PROVEN_JSON_NUMBER,
    string = c.PROVEN_JSON_STRING,
    array = c.PROVEN_JSON_ARRAY,
    object = c.PROVEN_JSON_OBJECT,
    invalid = c.PROVEN_JSON_INVALID,
};

/// Check if string is valid JSON via libproven.
pub fn isValid(input: []const u8) bool {
    const result = c.proven_json_is_valid(input.ptr, input.len);
    return result.status == c.PROVEN_OK and result.value;
}

/// Get JSON value type at root level via libproven.
pub fn getType(input: []const u8) JsonType {
    const result = c.proven_json_get_type(input.ptr, input.len);
    return @enumFromInt(result);
}

test "isValid" {
    try std.testing.expect(isValid("{}"));
    try std.testing.expect(isValid("{\"foo\": 42}"));
    try std.testing.expect(isValid("[1, 2, 3]"));
    try std.testing.expect(!isValid("{invalid}"));
}

test "getType" {
    try std.testing.expectEqual(JsonType.object, getType("{}"));
    try std.testing.expectEqual(JsonType.array, getType("[1,2]"));
    try std.testing.expectEqual(JsonType.string, getType("\"hello\""));
    try std.testing.expectEqual(JsonType.number, getType("42"));
    try std.testing.expectEqual(JsonType.invalid, getType("not json"));
}
